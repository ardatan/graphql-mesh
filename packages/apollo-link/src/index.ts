import { ApolloLink, FetchResult, Observable, Operation } from '@apollo/client/core';
import { MeshInstance } from '@graphql-mesh/runtime';
import { getOperationAST } from 'graphql';

export class MeshApolloLink extends ApolloLink {
  mesh$: Promise<MeshInstance>;
  constructor(getBuiltMesh: () => Promise<MeshInstance>) {
    super();
    this.mesh$ = getBuiltMesh();
  }

  request(operation: Operation): Observable<FetchResult> {
    const operationAst = getOperationAST(operation.query, operation.operationName);
    if (!operationAst) {
      throw new Error('GraphQL operation not found');
    }
    return new Observable(observer => {
      Promise.resolve()
        .then(async () => {
          const mesh = await this.mesh$;
          try {
            if (operationAst.operation === 'subscription') {
              const subscriptionResult = await mesh.subscribe(
                operation.query,
                operation.variables,
                operation.getContext(),
                {},
                operation.operationName
              );
              if (Symbol.asyncIterator in subscriptionResult) {
                for await (const result of subscriptionResult) {
                  if (observer.closed) {
                    return;
                  }
                  observer.next(result);
                }
                observer.complete();
              }
            } else {
              const result = await mesh.execute(
                operation.query,
                operation.variables,
                operation.getContext(),
                {},
                operation.operationName
              );

              if (!observer.closed) {
                observer.next(result);
                observer.complete();
              }
            }
          } catch (error) {
            if (!observer.closed) {
              observer.error(error);
            }
          }
        })
        .catch(error => {
          if (!observer.closed) {
            observer.error(error);
          }
        });
    });
  }
}
