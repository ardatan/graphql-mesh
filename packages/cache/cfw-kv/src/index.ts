import { KeyValueCache, Logger } from '@graphql-mesh/types';
import { KVNamespace } from '@cloudflare/workers-types';

export default class CFWorkerKVCache implements KeyValueCache {
  private kvNamespace?: KVNamespace;
  constructor(config: { namespace: string; logger: Logger }) {
    this.kvNamespace = globalThis[config.namespace];
    if (this.kvNamespace === undefined) {
      // We don't use mocks because they increase the bundle size.
      config.logger.warn(`Make sure KV Namespace: ${config.namespace} exists.`);
    }
  }

  get<T>(key: string): Promise<T | undefined> {
    return this.kvNamespace?.get(key, 'json');
  }

  async getKeysByPrefix(prefix: string): Promise<string[]> {
    const result = await this.kvNamespace?.list({
      prefix,
    });

    if (!result) {
      return [];
    }

    return result.keys.map(keyEntry => keyEntry.name);
  }

  set(key: string, value: any, options?: { ttl?: number }): Promise<void> {
    return this.kvNamespace?.put(key, JSON.stringify(value), {
      expirationTtl: options?.ttl,
    });
  }

  delete(key: string): Promise<void> {
    return this.kvNamespace?.delete(key);
  }
}
