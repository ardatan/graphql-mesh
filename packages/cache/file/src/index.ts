import { KeyValueCache } from '@graphql-mesh/types';
import { isAbsolute, join } from 'path';
import { get, set } from 'lodash';
import { existsSync } from 'fs';
import DataLoader from 'dataloader';
import { readJSONSync, writeJSON } from '@graphql-mesh/utils';

export default class FileCache<V = string> implements KeyValueCache<V> {
  json: Record<string, V>;
  absolutePath: string;
  writeDataLoader: DataLoader<any, any>;
  constructor({ path }: { path: string }) {
    this.absolutePath = isAbsolute(path) ? path : join(process.cwd(), path);
    this.json = existsSync(this.absolutePath) ? readJSONSync(this.absolutePath) : {};
    this.writeDataLoader = new DataLoader(async keys => {
      await writeJSON(this.absolutePath, this.json);
      return keys;
    });
  }

  async get(name: string) {
    return get(this.json, name);
  }

  async set(name: string, value: V) {
    set(this.json, name, value);
    this.writeDataLoader.load({});
  }

  async delete(name: string) {
    this.set(name, undefined);
  }
}
