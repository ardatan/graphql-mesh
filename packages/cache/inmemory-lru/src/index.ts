import LRUCache from 'lru-cache';
import { KeyValueCache, KeyValueCacheSetOptions } from '@graphql-mesh/types';

// Based on https://github.com/apollographql/apollo-server/blob/master/packages/apollo-datasource-rest/src/HTTPCache.ts

export class InMemoryLRUCache<V = string> implements KeyValueCache<V> {
  private store: LRUCache<string, V>;

  constructor({
    max = Infinity,
    length = (item: V) => (Array.isArray(item) || typeof item === 'string') ? item.length : 1,
    ...options
  }: LRUCache.Options<string, V> = {}) {
    this.store = new LRUCache({
      max,
      length,
      ...options,
    });
  }

  async get(key: string) {
    return this.store.get(key);
  }

  async set(key: string, value: V, options?: KeyValueCacheSetOptions) {
    const maxAge = options?.ttl && options?.ttl * 1000;
    this.store.set(key, value, maxAge!);
  }

  async delete(key: string) {
    this.store.del(key);
  }
}