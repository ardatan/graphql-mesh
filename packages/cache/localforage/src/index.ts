import localforage from 'localforage';
import { KeyValueCache, KeyValueCacheSetOptions, YamlConfig } from '@graphql-mesh/types';

export default class LocalforageCache<V = string> implements KeyValueCache<V> {
  constructor(config?: YamlConfig.LocalforageConfig) {
    const driverNames = config?.driver || ['INDEXEDDB', 'WEBSQL', 'LOCALSTORAGE'];
    const runtimeConfig = {
      ...config,
      driver: driverNames.map(driverName => localforage[driverName]),
    };
    localforage.config(runtimeConfig);
  }

  async get(key: string) {
    const expiresAt = await localforage.getItem<number>(`${key}.expiresAt`);
    if (expiresAt && Date.now() > expiresAt) {
      await localforage.removeItem(key);
    }
    return localforage.getItem<V>(key.toString());
  }

  async set(key: string, value: V, options?: KeyValueCacheSetOptions) {
    const jobs: Promise<any>[] = [localforage.setItem<V>(key, value)];
    if (options?.ttl) {
      jobs.push(localforage.setItem(`${key}.expiresAt`, Date.now() + options.ttl * 1000));
    }
    await Promise.all(jobs);
  }

  delete(key: string) {
    return localforage.removeItem(key);
  }
}
