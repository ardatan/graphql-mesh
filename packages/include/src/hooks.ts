/* eslint-disable import/no-nodejs-modules */
// ONLY FOR NODE. register with `node --import @graphql-mesh/include/hooks <your script>`

import fs from 'node:fs/promises';
import module from 'node:module';
import path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { createPathsMatcher, getTsconfig } from 'get-tsconfig';
import { transform, type Transform } from 'sucrase';

const isDebug = ['1', 'y', 'yes', 't', 'true'].includes(String(process.env.DEBUG));

function debug(msg: string) {
  if (isDebug) {
    process.stderr.write(`[${new Date().toISOString()}] HOOKS ${msg}\n`);
  }
}

// eslint-disable-next-line dot-notation
const resolveFilename: (path: string) => string = module['_resolveFilename'];

let packedDepsPath = '';

let pathsMatcher: ((specifier: string) => string[]) | undefined;

export interface InitializeData {
  /**
   * Packed deps will be checked first, and enforced if present, during module resolution.
   * This allows us to consistently use the same module instance even if multiple are installed by the user.
   */
  packedDepsPath?: string;
  /**
   * tsconfig search path for registering tsconfig paths.
   *
   * @default process.env.MESH_INCLUDE_TSCONFIG_SEARCH_PATH || 'tsconfig.json'
   */
  tsconfigSearchPath?: string;
}

export const initialize: module.InitializeHook<InitializeData> = (data = {}) => {
  if (data.packedDepsPath) {
    packedDepsPath = data.packedDepsPath;
    debug(`Packed dependencies available at "${packedDepsPath}"`);
  }
  const tsconfig = getTsconfig(
    undefined,
    data.tsconfigSearchPath || process.env.MESH_INCLUDE_TSCONFIG_SEARCH_PATH || 'tsconfig.json',
  );
  if (tsconfig) {
    debug(`tsconfig found at "${tsconfig.path}"`);
    pathsMatcher = createPathsMatcher(tsconfig);
  }
};

function fixSpecifier(specifier: string, context: module.ResolveHookContext) {
  if (path.sep === '\\') {
    if (context.parentURL != null && context.parentURL[1] === ':') {
      context.parentURL = pathToFileURL(context.parentURL.replaceAll('/', '\\')).toString();
    }
    if (specifier[1] === ':' && specifier[2] === '/') {
      specifier = specifier.replaceAll('/', '\\');
    }
    if (specifier.startsWith('file://')) {
      specifier = fileURLToPath(specifier);
    }
    if (!specifier.startsWith('.') && !specifier.startsWith('file:') && specifier[1] === ':') {
      specifier = pathToFileURL(specifier).toString();
    }
  }
  if (specifier.startsWith('node_modules/') || specifier.startsWith('node_modules\\')) {
    specifier = specifier
      .replace('node_modules/', '')
      .replace('node_modules\\', '')
      .replace(/\\/g, '/');
  }
  return specifier;
}

export const resolve: module.ResolveHook = async (specifier, context, nextResolve) => {
  specifier = fixSpecifier(specifier, context);

  if (specifier.startsWith('node:')) {
    return nextResolve(specifier, context);
  }
  if (module.builtinModules.includes(specifier)) {
    return nextResolve(specifier, context);
  }

  if (!specifier.startsWith('.') && packedDepsPath) {
    try {
      debug(
        `Trying packed dependency "${specifier}" for "${context.parentURL?.toString() || '.'}"`,
      );
      const resolved = resolveFilename(path.join(packedDepsPath, specifier));
      debug(`Possible packed dependency "${specifier}" to "${resolved}"`);
      return await nextResolve(fixSpecifier(resolved, context), context);
    } catch {
      // noop
    }
  }

  try {
    // debug(`Trying default resolve for "${specifier}"`);
    return await nextResolve(specifier, context);
  } catch (e) {
    try {
      debug(`Trying default resolve for "${specifier}" failed; trying alternatives`);
      const specifierWithoutJs = specifier.endsWith('.js') ? specifier.slice(0, -3) : specifier;
      const specifierWithTs = specifierWithoutJs + '.ts'; // TODO: .mts or .cts
      debug(`Trying "${specifierWithTs}"`);
      return await nextResolve(fixSpecifier(specifierWithTs, context), context);
    } catch (e) {
      try {
        return await nextResolve(fixSpecifier(resolveFilename(specifier), context), context);
      } catch {
        try {
          const specifierWithoutJs = specifier.endsWith('.js') ? specifier.slice(0, -3) : specifier;
          // usual filenames tried, could be a .ts file?
          return await nextResolve(
            fixSpecifier(
              resolveFilename(
                specifierWithoutJs + '.ts', // TODO: .mts or .cts?
              ),
              context,
            ),
            context,
          );
        } catch {
          // not a .ts file, try the tsconfig paths if available
          if (pathsMatcher) {
            for (const possiblePath of pathsMatcher(specifier)) {
              try {
                return await nextResolve(
                  fixSpecifier(resolveFilename(possiblePath), context),
                  context,
                );
              } catch {
                try {
                  const possiblePathWithoutJs = possiblePath.endsWith('.js')
                    ? possiblePath.slice(0, -3)
                    : possiblePath;
                  // the tsconfig path might point to a .ts file, try it too
                  return await nextResolve(
                    fixSpecifier(
                      resolveFilename(
                        possiblePathWithoutJs + '.ts', // TODO: .mts or .cts?
                      ),
                      context,
                    ),
                    context,
                  );
                } catch {
                  // noop
                }
              }
            }
          }
        }
      }
    }

    // none of the alternatives worked, fail with original error
    throw e;
  }
};

export const load: module.LoadHook = async (url, context, nextLoad) => {
  if (path.sep === '\\' && !url.startsWith('file:') && url[1] === ':') {
    debug(`Fixing Windows path at "${url}"`);
    url = `file:///${url.replace(/\\/g, '/')}`;
  }
  if (/\.(m|c)?ts$/.test(url)) {
    // debug(`Transpiling TypeScript file at "${url}"`);
    const filePath = fileURLToPath(url);
    let source: string;
    try {
      source = await fs.readFile(filePath, 'utf8');
    } catch (e) {
      throw new Error(`Failed to read file at "${url}"; ${e?.stack || e}`);
    }
    let format: 'module' | 'commonjs';
    if (/\.ts$/.test(url)) {
      // try {
      //   const { isSea } = await import('node:sea');
      //   format = isSea() ? 'commonjs' : 'module';
      // } catch {
      format = 'module';
      // }
    } else if (/\.mts$/.test(url)) {
      format = 'module';
    } else if (/\.cts$/.test(url)) {
      format = 'commonjs';
    }
    const transforms: Transform[] = ['typescript'];
    if (format === 'commonjs') {
      transforms.push('imports');
    }
    const { code } = transform(source, { transforms });
    return {
      format,
      source: code,
      shortCircuit: true,
    };
  }
  return nextLoad(url, context);
};
