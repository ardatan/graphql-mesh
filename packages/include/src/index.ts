// eslint-disable-next-line import/no-nodejs-modules
import Module from 'node:module';
import { createPathsMatcher, getTsconfig } from 'get-tsconfig';
import createJITI from 'jiti';

const jiti = createJITI(
  /**
   * We intentionally provide an empty string here and let jiti handle the base URL.
   *
   * This is because `import.meta.url` is not available in CJS (and cant even be in the syntax)
   * and `__filename` is not available in ESM.
   */
  '',
);

/**
 * Import a module, ESM or CJS at the provided {@link path}.
 *
 * If the included module has a "default" export, it will be returned instead.
 *
 * If the module at {@link path} is not found, `null` will be returned.
 */
export async function include<T = any>(path: string): Promise<T> {
  const module = await jiti.import(path, {});
  if (!module) {
    throw new Error('Included module is empty');
  }
  if (typeof module !== 'object') {
    throw new Error(`Included module is not an object, is instead "${typeof module}"`);
  }
  if ('default' in module) {
    return module.default as T;
  }
  return module as T;
}

export interface RegisterTsconfigPathsOptions {
  /**
   * Accepts a path to a file or directory to search up for a {@link configName} file.
   *
   * @default process.cwd()
   */
  cwd?: string;
  /**
   * The file name of the TypeScript config file.
   *
   * Can be changed using the `MESH_INCLUDE_TSCONFIG_NAME` environment variable.
   *
   * @default 'tsconfig.json'
   */
  configName?: string;
}

/**
 * Parses the closest `(t|j)sconfig.json` paths and augments Node's module
 * resolution to consider those paths during imports.
 *
 * @returns The unregister function.
 */
export function registerTsconfigPaths({
  cwd = process.cwd(),
  configName = process.env.MESH_INCLUDE_TSCONFIG_NAME || 'tsconfig.json',
}: RegisterTsconfigPathsOptions = {}): () => void {
  const tsconfig = getTsconfig(cwd, configName);
  if (!tsconfig) return; // no tsconfig, no register

  const pathsMatcher = createPathsMatcher(tsconfig);

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-expect-error
  const originalResolveFilename = Module._resolveFilename;
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-expect-error
  Module._resolveFilename = (...args) => {
    const [path, ...rest] = args;
    try {
      return originalResolveFilename(...args);
    } catch (e) {
      if (e.code !== 'MODULE_NOT_FOUND') {
        // throw non-module_not_found errors immediately
        throw e;
      }

      // original resolve didnt find the module, try resolving using tsconfig paths
      for (const possiblePath of pathsMatcher(path)) {
        try {
          return originalResolveFilename(possiblePath, ...rest);
        } catch {
          // noop
        }
      }

      // tsconfig paths couldnt resolve the module either, throw original error
      throw e;
    }
  };

  return function unregisterTsconfigPaths() {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    Module._resolveFilename = originalResolveFilename;
  };
}
