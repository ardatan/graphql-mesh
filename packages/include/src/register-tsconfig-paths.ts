import { registerTsconfigPaths } from './index.js';

registerTsconfigPaths();
