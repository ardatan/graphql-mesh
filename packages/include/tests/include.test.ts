import fs from 'node:fs';
import path from 'node:path';
import { globSync } from 'glob';
import { include } from '../src/index';

const fixtures = globSync(path.join(__dirname, 'fixtures', '*')).map(p => ({
  fixture: path.basename(p),
  cwd: p,
  pathToInclude: path.join(p, fs.readFileSync(path.join(p, 'include.txt'), 'utf8').trim()),
}));

it.each(fixtures)('should include in fixture $fixture', async ({ pathToInclude }) => {
  // TODO: cant test tsconfig paths because jest doesnt allow manipulating Module._resolveFilename
  // const unregister = registerTsconfigPaths({ cwd });
  // await using _ = {
  //   [Symbol.dispose]() {
  //     unregister();
  //   },
  // };
  await expect(include(pathToInclude)).resolves.toMatchSnapshot();
});
