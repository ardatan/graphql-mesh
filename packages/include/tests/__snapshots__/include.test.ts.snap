// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should include in fixture cjs-imports-another-no-ext 1`] = `
{
  "john": "doe",
}
`;

exports[`should include in fixture cjs-imports-esm-no-ext 1`] = `
{
  "john": "doe",
}
`;

exports[`should include in fixture imports-another-cjs-ext-cjs 1`] = `"doe"`;

exports[`should include in fixture imports-another-js-ext 1`] = `"doe"`;

exports[`should include in fixture imports-another-no-ext 1`] = `"doe"`;

exports[`should include in fixture imports-another-no-ext-cjs 1`] = `"doe"`;

exports[`should include in fixture js-imports-another-no-ext 1`] = `"doe"`;

exports[`should include in fixture simple-default-export 1`] = `
{
  "john": "doe",
}
`;

exports[`should include in fixture simple-named-export 1`] = `
{
  "john": "doe",
}
`;
